veps/sig-network/passt-migration-proposal.md

```diff
@@ -0,0 +1,333 @@
+It improves I/O performance through zero-copy mechanisms and reduced context-switching overhead. Additionally, it is required for seamless TCP migration.
+
+`vhost-user` must be enabled in the DomainXML. Small code changes are required in passt sidecar [code](https://github.com/kubevirt/kubevirt/tree/release-1.5/cmd/sidecars/network-passt-binding). The format is described [here](https://libvirt.org/formatdomain.html#vhost-user-connection-with-passt-backend).
+
+#### CNI plugin
+
+The [passt-cni](https://github.com/kubevirt/kubevirt/tree/release-1.5/cmd/cniplugins/passt-binding) process is called by `Multus` during primary network interface setup/removal. It runs two sysctl commands:
+1. Allow binding to all ports starting from 0.
+2. Set ping group range to virt-launcher's user ID.
+> [!NOTE] 
+> No changes are expected in the `passt` CNI code.
+
+### Build and Deployment
+
+#### Add passt-repair binary to virt-handler image
+
+The `passt-repair` binary will be added to the virt-handler container image installed from the `passt` RPM.
+To maintain storage efficiency and eliminate unnecessary security risks, the `passt-helper` binary will be extracted from the `passt` RPM and installed in the virt-handler container image without the `passt` binary.
+
+#### passt CNI plugin build and deployment manifests
+
+The [passt-binding directory](https://github.com/kubevirt/kubevirt/tree/release-1.5/cmd/cniplugins/passt-binding) currently includes a build system to create the CNI plugin binary.
+A few artifacts are required for the deployment of the network binding plugin:
+- A container image build and publish process for the sidecar container.
+- A static daemonset manifest which serves to deploy the CNI.
+- A static manifest for applying a `network-attachment-definition` CR named `netbindingpasst` referencing to the CNI.
+[CNAO](https://github.com/kubevirt/cluster-network-addons-operator) will be enhanced to store and automate the deployment process.
+
+## API Examples
+
+No API changes are expected in kubevirt/kubevirt.
+### passt network binding plugin configuration in KubeVirt CR
+```yaml
+apiVersion: kubevirt.io/v1
+kind: KubeVirt
+spec:
+  configuration:
+    network:
+      binding:
+        passt:
+          computeResourceOverhead:
+            requests:
+              memory: 250Mi
+          migration: {}
+          networkAttachmentDefinition: default/netbindingpass
+          sidecarImage: quay.io/kubevirt/network-passt-binding:<tag/sha>
+```
+
+### VM network interface with passt network binding 
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+spec:
+  template:
+    spec:
+      domain:
+        devices:
+          interfaces:
+          - name: passtnet
+            binding:
+              name: passt
+            ports:
+            - name: http
+              port: 80
+              protocol: TCP
+```
+
+## Alternatives
```

Comment:
@nirdothan Can you please tell if its possible as an alternative to run the passt-repair in a separate container in the virt-handler?

Reply:
> run the passt-repair in a separate container in the virt-handler?

...may I ask why?

Reply:
I can't easily answer that.  Wherever it runs, `passt-repair` needs to have `CAP_NET_ADMIN` in the user namespace which owns the network namespace of the virt-launcher pod.  I don't know enough about how k8s organises its namespaces to know if that's easy, tricky or outright impossible.

Reply:
@nirdothan Thank you! My main concern is the security aspect. I was  thinking that if we can containerize the `passt-repair` we would limit it with the capabilities it needs. Another aspect is stability, if passt-repair will crash it can have implications on the virt-handler.

Reply:
@nirdothan Sorry, yes the supply chain aspect is the one I am worried about. Basically you are asking to run a forked binary blob from the virt-handler, this is my concern as a maintainer - I don't see the source code of what I am running.

Reply:
@nirdothan Although we already doing it but we do it in virt-launcher which is unprivileged.

Reply:
I would be more relaxed if it `passt-repair` source would be part of the repo, or somehow part of the passt plugin. But maybe its just me.

Reply:
@sbrivio-rh Hi, by forked I meant the goroutine in which it will run. You are correct that its not feasible to control all the sw stack in one place. Currently I don't have any better alternative, I also had a talk with Nir and he mentioned Seitan, which can be a solution for the long term.

Reply:
> @sbrivio-rh Hi, by forked I meant the goroutine in which it will run.

Oh, I see. Well, yes, we can't do much better at the moment.

Right, [seitan](https://seitan.rocks/) is what I would have gone for in an ideal world, but it's not ready yet as we couldn't justify spending much effort on it in the recent past, and then when the need comes up, we have to settle for (and spend time on) minimal, relatively "quick" solutions such as passt-repair instead. :)

It's kind of a chicken-and-egg problem in some sense (fine print: seitan contains neither).